lib/tree.rb

```diff
@@ -16,48 +16,200 @@ def initialize
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree.
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])  
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, nodes_array)
+    unless current.nil?
+      inorder_helper(current.left, nodes_array)
+      nodes_array << { key: current.key, value: current.value }
+      inorder_helper(current.right, nodes_array)
+    end
+
+    return nodes_array
+  end
+
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree.
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  def preorder_helper(current, nodes_array)
+    unless current.nil?
+      nodes_array << { key: current.key, value: current.value }
+      preorder_helper(current.left, nodes_array)
+      preorder_helper(current.right, nodes_array)
+    end
+
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree. 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  def postorder_helper(current, nodes_array)
+    unless current.nil?
+      postorder_helper(current.left, nodes_array)
+      postorder_helper(current.right, nodes_array)
+      nodes_array << { key: current.key, value: current.value }
+    end
+
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree in order to determine the height of the tree
+  # Space Complexity: O(1) because the amount of data stored is constant, though the space required in memory for the stack trace will be O(n)
   def height
-    raise NotImplementedError
+    return 0 if @root.nil?
+
+    return height_helper(@root)
+  end
+
+  def height_helper(current)
+    return 0 if current.nil?
+
+    left = 1 + height_helper(current.left)
+    right = 1 + height_helper(current.right)
+
+    return [left, right].max
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree. 
   def bfs
```

Comment:
Unfortunately this is more like a preorder traversal, this is because you traverse the entire left subtree before you traverse the right subtree.  Remember all nodes at level 2 should be next to each other in the array and level 3 etc.